In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("cust_seg.csv")

C:\Users\rayng\AppData\Local\Temp\ipykernel_9268\1948748220.py:1: DtypeWarning: Columns (16) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("cust_seg.csv")


In [3]:
# What kind of data does each column hold?

for col in df:
    print(col + ":", df[col].unique())

Unnamed: 0: [     0      1      2 ... 999997 999998 999999]
fecha_dato: ['2015-01-28' '2015-02-28']
ncodpers: [1375586 1050611 1050612 ... 1149999 1150908 1183305]
ind_empleado: ['N' nan 'A' 'B' 'F' 'S']
pais_residencia: ['ES' nan 'CA' 'CH' 'CL' 'IE' 'AT' 'NL' 'FR' 'GB' 'DE' 'DO' 'BE' 'AR' 'VE'
 'US' 'MX' 'BR' 'IT' 'EC' 'PE' 'CO' 'HN' 'FI' 'SE' 'AL' 'PT' 'MZ' 'CN'
 'TW' 'PL' 'IN' 'CR' 'NI' 'HK' 'AD' 'CZ' 'AE' 'MA' 'GR' 'PR' 'RO' 'IL'
 'RU' 'GT' 'GA' 'NO' 'SN' 'MR' 'UA' 'BG' 'PY' 'EE' 'SV' 'ET' 'CM' 'SA'
 'CI' 'QA' 'LU' 'PA' 'BA' 'BO' 'AU' 'BY' 'KE' 'SG' 'HR' 'MD' 'SK' 'TR'
 'AO' 'CU' 'GQ' 'EG' 'ZA' 'DK' 'UY' 'GE' 'TH' 'DZ' 'LB' 'JP' 'NG' 'PK'
 'TN' 'TG' 'KR' 'GH' 'RS' 'VN' 'PH' 'KW' 'NZ' 'MM' 'KH' 'GI' 'SL' 'GN'
 'GW' 'OM' 'CG' 'LV' 'LT' 'ML' 'MK' 'HU' 'IS' 'LY' 'CF' 'GM' 'KZ' 'CD'
 'BZ']
sexo: ['H' 'V' nan]
age: [' 35' ' 23' ' 22' ' 24' ' 65' ' 28' ' 25' ' 26' ' 53' ' 27' ' 32' ' 37'
 ' 31' ' 39' ' 63' ' 33' ' 55' ' 42' ' 58' ' 38' ' 50' ' 30' ' 45' ' 44'
 ' 36' ' 29' ' 60' ' 57' ' 67

In [4]:
# Looking at the code above, it is appropriate to remove all trailing spaces from the "age" and "antiguedad" columns. There is a method named strip to do this

df["age"] = df["age"].apply(lambda string: string.strip())
df["antiguedad"] = df["antiguedad"].apply(lambda string: string.strip())

In [5]:
# Convert "NA" strings into the Pandas nan type

df["age"] = df["age"].replace("NA", np.nan)
df["antiguedad"] = df["antiguedad"].replace("NA", np.nan)

In [6]:
# The "age" and "antiguedad" columns are encoded as strings, not numbers

def convert_to_int(data):
    try:
        return int(data)
    except:
        return data

df["age"] = df["age"].apply(convert_to_int)
df["antiguedad"] = df["antiguedad"].apply(convert_to_int)

In [7]:
# What kind of data does each column hold?

for col in df:
    print(col + ":", df[col].unique())

Unnamed: 0: [     0      1      2 ... 999997 999998 999999]
fecha_dato: ['2015-01-28' '2015-02-28']
ncodpers: [1375586 1050611 1050612 ... 1149999 1150908 1183305]
ind_empleado: ['N' nan 'A' 'B' 'F' 'S']
pais_residencia: ['ES' nan 'CA' 'CH' 'CL' 'IE' 'AT' 'NL' 'FR' 'GB' 'DE' 'DO' 'BE' 'AR' 'VE'
 'US' 'MX' 'BR' 'IT' 'EC' 'PE' 'CO' 'HN' 'FI' 'SE' 'AL' 'PT' 'MZ' 'CN'
 'TW' 'PL' 'IN' 'CR' 'NI' 'HK' 'AD' 'CZ' 'AE' 'MA' 'GR' 'PR' 'RO' 'IL'
 'RU' 'GT' 'GA' 'NO' 'SN' 'MR' 'UA' 'BG' 'PY' 'EE' 'SV' 'ET' 'CM' 'SA'
 'CI' 'QA' 'LU' 'PA' 'BA' 'BO' 'AU' 'BY' 'KE' 'SG' 'HR' 'MD' 'SK' 'TR'
 'AO' 'CU' 'GQ' 'EG' 'ZA' 'DK' 'UY' 'GE' 'TH' 'DZ' 'LB' 'JP' 'NG' 'PK'
 'TN' 'TG' 'KR' 'GH' 'RS' 'VN' 'PH' 'KW' 'NZ' 'MM' 'KH' 'GI' 'SL' 'GN'
 'GW' 'OM' 'CG' 'LV' 'LT' 'ML' 'MK' 'HU' 'IS' 'LY' 'CF' 'GM' 'KZ' 'CD'
 'BZ']
sexo: ['H' 'V' nan]
age: [ 35.  23.  22.  24.  65.  28.  25.  26.  53.  27.  32.  37.  31.  39.
  63.  33.  55.  42.  58.  38.  50.  30.  45.  44.  36.  29.  60.  57.
  67.  47.  nan  34.  48.  46.  5

Here we shall investigate the NA in the data.

In [8]:
# How many NA are in each column?
for col in df:
    print(col + ":", df[col].isna().sum())

Unnamed: 0: 0
fecha_dato: 0
ncodpers: 0
ind_empleado: 10782
pais_residencia: 10782
sexo: 10786
age: 10782
fecha_alta: 10782
ind_nuevo: 10782
antiguedad: 10782
indrel: 10782
ult_fec_cli_1t: 998899
indrel_1mes: 10782
tiprel_1mes: 10782
indresi: 10782
indext: 10782
conyuemp: 999822
canal_entrada: 10861
indfall: 10782
tipodom: 10782
cod_prov: 17734
nomprov: 17734
ind_actividad_cliente: 10782
renta: 175183
ind_ahor_fin_ult1: 0
ind_aval_fin_ult1: 0
ind_cco_fin_ult1: 0
ind_cder_fin_ult1: 0
ind_cno_fin_ult1: 0
ind_ctju_fin_ult1: 0
ind_ctma_fin_ult1: 0
ind_ctop_fin_ult1: 0
ind_ctpp_fin_ult1: 0
ind_deco_fin_ult1: 0
ind_deme_fin_ult1: 0
ind_dela_fin_ult1: 0
ind_ecue_fin_ult1: 0
ind_fond_fin_ult1: 0
ind_hip_fin_ult1: 0
ind_plan_fin_ult1: 0
ind_pres_fin_ult1: 0
ind_reca_fin_ult1: 0
ind_tjcr_fin_ult1: 0
ind_valo_fin_ult1: 0
ind_viv_fin_ult1: 0
ind_nomina_ult1: 5402
ind_nom_pens_ult1: 5402
ind_recibo_ult1: 0


In [9]:
df[df["ind_empleado"].isna()]

Unnamed: 0  fecha_dato  ncodpers ind_empleado pais_residencia sexo  \
261            261  2015-01-28   1050741          NaN             NaN  NaN   
1029          1029  2015-01-28   1051017          NaN             NaN  NaN   
1063          1063  2015-01-28   1051064          NaN             NaN  NaN   
1154          1154  2015-01-28   1051387          NaN             NaN  NaN   
1779          1779  2015-01-28   1048660          NaN             NaN  NaN   
...            ...         ...       ...          ...             ...  ...   
998780      998780  2015-02-28   1148591          NaN             NaN  NaN   
999111      999111  2015-02-28   1148971          NaN             NaN  NaN   
999115      999115  2015-02-28   1148968          NaN             NaN  NaN   
999577      999577  2015-02-28   1147591          NaN             NaN  NaN   
999724      999724  2015-02-28   1148181          NaN             NaN  NaN   

        age fecha_alta  ind_nuevo  antiguedad  ...  ind_hip_fin_ult1  \
261     NaN        NaN        NaN         NaN  ...                 0   
1029    NaN        NaN        NaN         NaN  ...                 0   
1063    NaN        NaN        NaN         NaN  ...                 0   
1154    NaN        NaN        NaN         NaN  ...                 0   
1779    NaN        NaN        NaN         NaN  ...                 0   
...     ...        ...        ...         ...  ...               ...   
998780  NaN        NaN        NaN         NaN  ...                 0   
999111  NaN        NaN        NaN         NaN  ...                 0   
999115  NaN        NaN        NaN         NaN  ...                 0   
999577  NaN        NaN        NaN         NaN  ...                 0   
999724  NaN        NaN        NaN         NaN  ...                 0   

       ind_plan_fin_ult1  ind_pres_fin_ult1 ind_reca_fin_ult1  \
261                    0                  0                 0   
1029                   0                  0                 0   
1063                   0                  0                 0   
1154                   0                  0                 0   
1779                   0                  0                 0   
...                  ...                ...               ...   
998780                 0                  0                 0   
999111                 0                  0                 0   
999115                 0                  0                 0   
999577                 0                  0                 0   
999724                 0                  0                 0   

       ind_tjcr_fin_ult1 ind_valo_fin_ult1 ind_viv_fin_ult1 ind_nomina_ult1  \
261                    0                 0                0             0.0   
1029                   0                 0                0             0.0   
1063                   0                 0                0             0.0   
1154                   0                 0                0             0.0   
1779                   0                 0                0             0.0   
...                  ...               ...              ...             ...   
998780                 0                 0                0             0.0   
999111                 0                 0                0             NaN   
999115                 0                 0                0             0.0   
999577                 0                 0                0             NaN   
999724                 0                 0                0             0.0   

       ind_nom_pens_ult1  ind_recibo_ult1  
261                  0.0                0  
1029                 0.0                0  
1063                 0.0                0  
1154                 0.0                0  
1779                 0.0                0  
...                  ...              ...  
998780               0.0                0  
999111               NaN                0  
999115               0.0                0  
999577               NaN            

In [10]:
# How many NA values are there per customer?

df["unrecorded_values"] = df.isna().sum(axis=1)
df["unrecorded_values"]

0         2
1         2
2         2
3         2
4         3
         ..
999995    2
999996    2
999997    2
999998    2
999999    2
Name: unrecorded_values, Length: 1000000, dtype: int64

In [11]:
# Which customers have the most NA values?

df["unrecorded_values"].value_counts()

2     825602
3     156380
5       6972
21      5480
23      5302
1        175
4         84
6          5
Name: unrecorded_values, dtype: int64

In [12]:
# The result above shows that about 10782 customers have 21 or 23 missing values - why were these info not recorded? Which columns are commonly not recorded for these customers?

df[df["unrecorded_values"] == 21].isna().sum(axis=0)

Unnamed: 0                  0
fecha_dato                  0
ncodpers                    0
ind_empleado             5480
pais_residencia          5480
sexo                     5480
age                      5480
fecha_alta               5480
ind_nuevo                5480
antiguedad               5480
indrel                   5480
ult_fec_cli_1t           5480
indrel_1mes              5480
tiprel_1mes              5480
indresi                  5480
indext                   5480
conyuemp                 5480
canal_entrada            5480
indfall                  5480
tipodom                  5480
cod_prov                 5480
nomprov                  5480
ind_actividad_cliente    5480
renta                    5480
ind_ahor_fin_ult1           0
ind_aval_fin_ult1           0
ind_cco_fin_ult1            0
ind_cder_fin_ult1           0
ind_cno_fin_ult1            0
ind_ctju_fin_ult1           0
ind_ctma_fin_ult1           0
ind_ctop_fin_ult1           0
ind_ctpp_fin_ult1           0
ind_deco_f

In [13]:
df[df["unrecorded_values"] == 23].isna().sum(axis=0)

Unnamed: 0                  0
fecha_dato                  0
ncodpers                    0
ind_empleado             5302
pais_residencia          5302
sexo                     5302
age                      5302
fecha_alta               5302
ind_nuevo                5302
antiguedad               5302
indrel                   5302
ult_fec_cli_1t           5302
indrel_1mes              5302
tiprel_1mes              5302
indresi                  5302
indext                   5302
conyuemp                 5302
canal_entrada            5302
indfall                  5302
tipodom                  5302
cod_prov                 5302
nomprov                  5302
ind_actividad_cliente    5302
renta                    5302
ind_ahor_fin_ult1           0
ind_aval_fin_ult1           0
ind_cco_fin_ult1            0
ind_cder_fin_ult1           0
ind_cno_fin_ult1            0
ind_ctju_fin_ult1           0
ind_ctma_fin_ult1           0
ind_ctop_fin_ult1           0
ind_ctpp_fin_ult1           0
ind_deco_f

In 5480 customers, the following columns are not recorded:

* ind_empleado
* pais_residencia
* sexo
* age
* fecha_alta
* ind_nuevo
* antiguedad
* indrel
* ult_fec_cli_1t
* indrel_1mes
* tiprel_1mes
* indresi
* indext
* conyuemp
* canal_entrada
* indfall
* tipodom
* cod_prov
* nomprov
* ind_actividad_cliente
* renta

In 5302 customers, in addition to the above 21 features, the features ind_nomina_ult1 and ind_nom_pens_ult1 are unrecorded.

Notice that there is some strange data that is incorrect. For instance, the antiguedad (customer seniority) column includes a value of -999999. This is probably a placeholder for an unknown value, so we could replace this with nan.

In [14]:
df["antiguedad"] = df["antiguedad"].replace(-999999, np.nan)

What to do with the NA values? Where the data is quantitative and continuous, we may assume the average or the median value to fill in the blanks. Where the data is categorical (even if numerically encoded), it will depend on the context. First, let's deal with the quantitative fields:

* age
* fecha_alta
* ult_fec_cli_1t
* antiguedad
* renta

Of these, only age is a continuous variable, making it appropriate to use the mean. For the rest, we shall use the median.

In [15]:
for col in ["renta"]:
    df[col] = df[col].fillna(df[col].mean())

In [16]:
for col in ["age", "antiguedad"]:
    df[col] = df[col].fillna(df[col].median())

In [17]:
from datetime import datetime

def convert_date(string):
    try:
        return datetime.strptime(string, "%Y-%m-%d").timestamp()
    except:
        return string

df["fecha_alta"] = df["fecha_alta"].fillna(datetime.fromtimestamp(df["fecha_alta"].map(convert_date).median()).strftime("%Y-%m-%d"))
df["ult_fec_cli_1t"] = df["ult_fec_cli_1t"].fillna(datetime.fromtimestamp(df["ult_fec_cli_1t"].map(convert_date).median()).strftime("%Y-%m-%d"))

For the other fields:

* ind_empleado
* pais_residencia
* sexo
* ind_nuevo
* indrel
* indrel_1mes
* tiprel_1mes
* indresi
* indext
* conyuemp
* canal_entrada
* indfall
* tipodom
* cod_prov
* nomprov
* ind_actividad_cliente

We can use the mode since they are all categorical.

In [43]:
for col in df:
    df[col] = df[col].fillna(df[col].mode()[0])

In [44]:
# How many NA are in each column?
for col in df:
    print(col + ":", df[col].isna().sum())

Unnamed: 0: 0
fecha_dato: 0
ncodpers: 0
ind_empleado: 0
pais_residencia: 0
sexo: 0
age: 0
fecha_alta: 0
ind_nuevo: 0
antiguedad: 0
indrel: 0
ult_fec_cli_1t: 0
indrel_1mes: 0
tiprel_1mes: 0
indresi: 0
indext: 0
conyuemp: 0
canal_entrada: 0
indfall: 0
tipodom: 0
cod_prov: 0
nomprov: 0
ind_actividad_cliente: 0
renta: 0
ind_ahor_fin_ult1: 0
ind_aval_fin_ult1: 0
ind_cco_fin_ult1: 0
ind_cder_fin_ult1: 0
ind_cno_fin_ult1: 0
ind_ctju_fin_ult1: 0
ind_ctma_fin_ult1: 0
ind_ctop_fin_ult1: 0
ind_ctpp_fin_ult1: 0
ind_deco_fin_ult1: 0
ind_deme_fin_ult1: 0
ind_dela_fin_ult1: 0
ind_ecue_fin_ult1: 0
ind_fond_fin_ult1: 0
ind_hip_fin_ult1: 0
ind_plan_fin_ult1: 0
ind_pres_fin_ult1: 0
ind_reca_fin_ult1: 0
ind_tjcr_fin_ult1: 0
ind_valo_fin_ult1: 0
ind_viv_fin_ult1: 0
ind_nomina_ult1: 0
ind_nom_pens_ult1: 0
ind_recibo_ult1: 0
unrecorded_values: 0


All NA have now been processed and replaced with an appropriate value. Now time to extract the transformed dataset.

In [57]:
df.drop(["Unnamed: 0", "unrecorded_values"], axis=1).to_csv("cust_seg_transformed.csv")